File: 0954-maximum-sum-circular-subarray/0954-maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        currentMax = -inf
        maxFromStart = [-inf for _ in range(len(nums))]
        maxFromEnd = [-inf for _ in range(len(nums))]
        last = 0
        currentSum = 0
        best = -inf
        for i, n in enumerate(nums):
            currentSum += n
            maxFromStart[i] = max(maxFromStart[i - 1], currentSum) if i - 1 >= 0 else currentSum
            currentMax = max(n, currentMax + n)
            best = max(currentMax, best)

        currentSum = 0
        maxFromEnd = -inf

        for i in range(len(nums) - 1, 0, -1):
            currentSum += nums[i]
            maxFromEnd = max(maxFromEnd, currentSum)
            
            best = max(
                maxFromStart[i - 1] + maxFromEnd,
                best
            )


        return best

File: 0954-maximum-sum-circular-subarray/0954-maximum-sum-circular-subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        long sumLargest = 0, sumSmallest = 0, total = 0, smallest = LONG_MAX, largest = LONG_MIN;

        for (int i = 0; i < nums.size(); i++) {
            long n = nums[i];
            sumLargest = max(n, sumLargest + n);
            largest = max(sumLargest, largest);
            sumSmallest = min(n, sumSmallest + n);
            smallest = min(sumSmallest, smallest);
            total += n;
        }

        return max(
            largest,
            total != smallest ? total - smallest : LONG_MIN
        );
    }
};